pvlib/singlediode_methods.py

```diff
@@ -0,0 +1,347 @@
+                brentq(fv, 0.0, voc, args=(v, iph, isat, rs, rsh, gamma))
+        )
+        vd = vec_fun(voc_est, voltage, *args)
+    elif method.lower() == 'newton':
+        # make sure all args are numpy arrays if max size > 1
+        size, shape = _get_size_and_shape((voltage,) + args)
+        if size > 1:
+            args = [np.asarray(arg) for arg in args]
+        # newton uses initial guess for the output shape
+        # copy v0 to a new array and broadcast it to the shape of max size
+        if shape is not None:
+            v0 = np.broadcast_to(voltage, shape).copy()
+        else:
+            v0 = voltage
+        # x0 and x in func are the same reference! DO NOT set x0 to voltage!
+        # voltage in fv(x, voltage, *a) MUST BE CONSTANT!
+        vd = newton(func=lambda x, *a: fv(x, voltage, *a), x0=v0,
+                    fprime=lambda x, *a: bishop88(x, *a, gradients=True)[4],
+                    args=args)
+    else:
+        raise NotImplementedError("Method '%s' isn't implemented" % method)
+    return bishop88(vd, *args)[0]
+
+
+def bishop88_v_from_i(current, photocurrent, saturation_current,
+                      resistance_series, resistance_shunt, nNsVth,
+                      method='newton'):
+    """
+    Find voltage given any current.
+
+    Parameters
+    ----------
+    current : numeric
+        current (I) in amperes [A]
+    photocurrent : numeric
+        photogenerated current (Iph or IL) in amperes [A]
+    saturation_current : numeric
+        diode dark or saturation current (Io or Isat) in amperes [A]
+    resistance_series : numeric
+        series resistance (Rs) in ohms
+    resistance_shunt : numeric
+        shunt resistance (Rsh) in ohms
+    nNsVth : numeric
+        product of diode ideality factor (n), number of series cells (Ns), and
+        thermal voltage (Vth = k_b * T / q_e) in volts [V]
+    method : str
+        one of two optional search methods: either ``'brentq'``, a reliable and
+        bounded method or ``'newton'`` which is the default.
+
+    Returns
+    -------
+    voltage : numeric
+        voltage (V) at the specified current (I) in volts [V]
+    """
+    # collect args
+    args = (photocurrent, saturation_current, resistance_series,
+            resistance_shunt, nNsVth)
+    # first bound the search using voc
+    voc_est = estimate_voc(photocurrent, saturation_current, nNsVth)
+
+    def fi(x, i, *a):
+        # calculate current residual given diode voltage "x"
+        return bishop88(x, *a)[0] - i
+
+    if method.lower() == 'brentq':
+        if brentq is NotImplemented:
+            raise ImportError('This function requires scipy')
+        # break out arguments for numpy.vectorize to handle broadcasting
+        vec_fun = np.vectorize(
+            lambda voc, i, iph, isat, rs, rsh, gamma:
+                brentq(fi, 0.0, voc, args=(i, iph, isat, rs, rsh, gamma))
+        )
+        vd = vec_fun(voc_est, current, *args)
```

Comment:
consider the same pattern as suggested in ``bishop88_i_from_v``.

Reply:
✔️